diff --git a/services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java b/services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java
index f9b925888c..af87bfa401 100755
--- a/services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java
+++ b/services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java
@@ -99,7 +99,7 @@ public class DefaultAuthenticationFlow implements AuthenticationFlow {
             if (inputData.containsKey("tryAnotherWay")) {
                 logger.trace("User clicked on link 'Try Another Way'");
 
-                processor.getAuthenticationSession().setAuthNote(AuthenticationProcessor.AUTHENTICATION_SELECTOR_SCREEN_DISPLAYED, "true");
+                processor.getAuthenticationSession().setAuthNote(AuthenticationProcessor.AUTHENTICATION_SELECTOR_SCREEN_DISPLAYED, model.getId());
                 return createSelectAuthenticatorsScreen(model);
             }
 
@@ -243,15 +243,17 @@ public class DefaultAuthenticationFlow implements AuthenticationFlow {
     @Override
     public Response processFlow() {
         logger.debugf("processFlow: %s", flow.getAlias());
-
-        if (Boolean.parseBoolean(processor.getAuthenticationSession().getAuthNote(AuthenticationProcessor.AUTHENTICATION_SELECTOR_SCREEN_DISPLAYED))) {
-            logger.tracef("Refreshed page on authentication selector screen");
+        String selector = processor.getAuthenticationSession().getAuthNote(AuthenticationProcessor.AUTHENTICATION_SELECTOR_SCREEN_DISPLAYED);
+        if (selector != null) {
             String lastExecutionId = processor.getAuthenticationSession().getAuthNote(AuthenticationProcessor.CURRENT_AUTHENTICATION_EXECUTION);
-            if (lastExecutionId != null) {
+            if (selector.equalsIgnoreCase(lastExecutionId)) {
+                logger.tracef("Refreshed page on authentication selector screen");
                 AuthenticationExecutionModel executionModel = processor.getRealm().getAuthenticationExecutionById(lastExecutionId);
                 if (executionModel != null) {
                     return createSelectAuthenticatorsScreen(executionModel);
                 }
+            } else {
+                processor.getAuthenticationSession().removeAuthNote(AuthenticationProcessor.AUTHENTICATION_SELECTOR_SCREEN_DISPLAYED);
             }
         }
 
diff --git a/services/src/main/java/org/keycloak/authentication/authenticators/resetcred/ResetCredentialEmail.java b/services/src/main/java/org/keycloak/authentication/authenticators/resetcred/ResetCredentialEmail.java
index 76b2487c8b..dddc005399 100755
--- a/services/src/main/java/org/keycloak/authentication/authenticators/resetcred/ResetCredentialEmail.java
+++ b/services/src/main/java/org/keycloak/authentication/authenticators/resetcred/ResetCredentialEmail.java
@@ -150,8 +150,15 @@ public class ResetCredentialEmail implements Authenticator, AuthenticatorFactory
 
     @Override
     public void action(AuthenticationFlowContext context) {
-        context.getUser().setEmailVerified(true);
-        context.success();
+        UserModel user = context.getUser();
+        String actionTokenUserId = context.getAuthenticationSession().getAuthNote(DefaultActionTokenKey.ACTION_TOKEN_USER_ID);
+        if (user != null && user.getId().equals(actionTokenUserId)) {
+            context.getUser().setEmailVerified(true);
+            context.success();
+        } else {
+            // not called via action token or any other weird situation
+            context.failure(AuthenticationFlowError.INVALID_USER);
+        }
     }
 
     @Override
diff --git a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetPasswordTest.java b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetPasswordTest.java
index 3288a622d8..c358bc266c 100644
--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetPasswordTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetPasswordTest.java
@@ -71,6 +71,7 @@ import org.keycloak.testsuite.pages.LoginPasswordResetPage;
 import org.keycloak.testsuite.pages.LoginPasswordUpdatePage;
 import org.keycloak.testsuite.pages.LogoutConfirmPage;
 import org.keycloak.testsuite.pages.PageUtils;
+import org.keycloak.testsuite.pages.SelectAuthenticatorPage;
 import org.keycloak.testsuite.updaters.ClientAttributeUpdater;
 import org.keycloak.testsuite.util.AccountHelper;
 import org.keycloak.testsuite.util.BrowserTabUtil;
@@ -97,6 +98,7 @@ import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.openqa.selenium.By;
+import org.openqa.selenium.JavascriptExecutor;
 import org.openqa.selenium.WebDriver;
 import org.openqa.selenium.WebElement;
 import org.openqa.selenium.firefox.FirefoxDriver;
@@ -123,6 +125,9 @@ public class ResetPasswordTest extends AbstractTestRealmKeycloakTest {
     @Rule
     public InfinispanTestTimeServiceRule ispnTestTimeService = new InfinispanTestTimeServiceRule(this);
 
+    @Page
+    SelectAuthenticatorPage selectAuthenticatorPage;
+
     @Drone
     @SecondBrowser
     protected WebDriver driver2;
@@ -1633,4 +1638,40 @@ public class ResetPasswordTest extends AbstractTestRealmKeycloakTest {
             appPage.assertCurrent();
         }
     }
+
+    private void submitForm(String action, String name, String value) {
+        ((JavascriptExecutor) driver).executeScript(
+                "var form = document.createElement('form');" +
+                "document.body.appendChild(form);" +
+                "form.method = 'POST';" +
+                "form.action ='"+ action + "';" +
+                "var element1 = document.createElement('input');" +
+                "element1.name='" + name +"';" +
+                "element1.value='" + value + "';" +
+                "form.appendChild(element1);" +
+                "form.submit();");
+    }
+
+    @Test
+    public void resetPasswordTryAnotherWay() throws Exception {
+        oauth.openLoginForm();
+        loginPage.resetPassword();
+        resetPasswordPage.assertCurrent();
+        String resetPasswordUrl = driver.getCurrentUrl();
+
+        // force a tryAnotherWay to set AUTHENTICATION_SELECTOR_SCREEN_DISPLAYED
+        submitForm(resetPasswordPage.getFormUrl(), "tryAnotherWay", "on");
+        selectAuthenticatorPage.assertCurrent();
+
+        // now execute the email action to send the email
+        WebElement form = driver.findElement(By.id("kc-select-credential-form"));
+        submitForm(form.getDomAttribute("action"), "username", "login-test");
+        loginPage.assertCurrent();
+        assertEquals("You should receive an email shortly with further instructions.", loginPage.getSuccessMessage());
+
+        // back to reset credential and check we are still waiting the email
+        driver.navigate().to(resetPasswordUrl);
+        loginPage.assertCurrent();
+        assertEquals("You should receive an email shortly with further instructions.", loginPage.getSuccessMessage());
+    }
 }
